lib/ts/Lexicon.h

```diff
@@ -0,0 +1,537 @@
+   * Caution must be used because the returned name must not leak and must be thread safe. The most
+   * common use would be for logging bad values.
+   */
+  self_type &set_default(const UnknownValueHandler &handler);
+
+protected:
+  // Because std::variant is broken up through clang 6, we have to do something uglier.
+  //  using NameDefault  = std::variant<std::monostate, std::string_view, UnknownValueHandler>;
+  //  using ValueDefault = std::variant<std::monostate, E, UnknownNameHandler>;
+
+  enum class Content {
+    NIL,    ///< Nothing, not set.
+    SCALAR, ///< A specific value/name.
+    HANDLER ///< A function
+  };
+
+  struct NilValue {
+  };
+
+  struct NameDefault {
+    using self_type = NameDefault;
+
+    NameDefault() = default;
+    ~NameDefault();
+
+    self_type &operator=(std::string_view name);
+    self_type &operator=(const UnknownValueHandler &handler);
+
+    std::string_view operator()(E value);
+
+    self_type &destroy();
+
+    Content _content{Content::NIL};
+    static constexpr size_t N = std::max<size_t>(sizeof(std::string_view), sizeof(UnknownValueHandler));
+    char _store[N];
+  };
+
+  struct ValueDefault {
+    using self_type = ValueDefault;
+
+    ValueDefault() = default;
+    ~ValueDefault();
+
+    self_type &operator=(E value);
+    self_type &operator=(const UnknownNameHandler &handler);
+
+    E operator()(std::string_view name);
+
+    self_type &destroy();
+
+    Content _content{Content::NIL};
+    static constexpr size_t N = std::max<size_t>(sizeof(E), sizeof(UnknownNameHandler));
+    char _store[N];
+  };
+
+  /// Each unique pair of value and name is stored as an instance of this class.
+  /// The primary is stored first and is therefore found by normal lookup.
+  struct Item {
+    Item(E, std::string_view);
+
+    E _value;               ///< Definition value.
+    std::string_view _name; ///< Definition name
+
+    /// Intrusive linkage for name lookup.
+    struct NameLinkage {
+      Item *_next{nullptr};
+      Item *_prev{nullptr};
+
+      static Item *&next_ptr(Item *);
+
+      static Item *&prev_ptr(Item *);
+
+      static std::string_view key_of(Item *);
+
+      static uint32_t hash_of(std::string_view s);
+
+      static bool equal(std::string_view const &lhs, std::string_view const &rhs);
+    } _name_link;
+
+    /// Intrusive linkage for value lookup.
+    struct ValueLinkage {
+      Item *_next{nullptr};
+      Item *_prev{nullptr};
+
+      static Item *&next_ptr(Item *);
+
+      static Item *&prev_ptr(Item *);
+
+      static E key_of(Item *);
+
+      static uintmax_t hash_of(E);
+
+      static bool equal(E lhs, E rhs);
+    } _value_link;
+  };
+
+  /// Copy @a name in to local storage.
+  std::string_view localize(std::string_view name);
+
+  /// Storage for names.
+  MemArena _arena{1024};
+  /// Access by name.
+  IntrusiveHashMap<typename Item::NameLinkage> _by_name;
```

Comment:
Yes, but I don't agree that in this case it would be simpler, in terms of the implementation, in `Lexicon`. The mapping also gives us nice locality of equivalent keys.